src/protocols/rdp/channels/rdpdr/rdpdr.h

```diff
@@ -0,0 +1,250 @@
+ *
+ * @param device
+ *     The guac_rdpdr_device of the device being freed.
+ */
+typedef void guac_rdpdr_device_free_handler(guac_rdp_common_svc* svc,
+        guac_rdpdr_device* device);
+
+struct guac_rdpdr_device {
+
+    /**
+     * The ID assigned to this device by the RDPDR plugin.
+     */
+    int device_id;
+
+    /**
+     * Device name, used for logging and for passthrough to the
+     * server.
+     */
+    const char* device_name;
+
+    /**
+     * The type of RDPDR device that this represents.
+     */
+    uint32_t device_type;
+
+    /**
+     * The DOS name of the device.  Max 8 bytes, including terminator.
+     */
+    const char *dos_name;
+    
+    /**
+     * The stream that stores the RDPDR device announcement for this device.
+     */
+    wStream* device_announce;
+    
+    /**
+     * The length of the device_announce wStream.
+     */
+    int device_announce_len;
+
+    /**
+     * Handler which should be called for every I/O request received.
+     */
+    guac_rdpdr_device_iorequest_handler* iorequest_handler;
+
+    /**
+     * Handler which should be called when the device is being freed.
+     */
+    guac_rdpdr_device_free_handler* free_handler;
+
+    /**
+     * Arbitrary data, used internally by the handlers for this device.
+     */
+    void* data;
+
+};
+
+/**
+ * Structure representing the current state of the Guacamole RDPDR plugin for
+ * FreeRDP.
+ */
+typedef struct guac_rdpdr {
+
+    /**
+     * The number of devices registered within the devices array.
+     */
+    int devices_registered;
+
+    /**
+     * Array of registered devices.
+     */
+    guac_rdpdr_device devices[8];
+
+} guac_rdpdr;
+
+/**
+ * Creates a new stream which contains the common DR_DEVICE_IOCOMPLETION header
+ * used for virtually all responses. Depending on the specific I/O completion
+ * being sent, additional space may be reserved within the resulting stream for
+ * additional fields. See:
+ *
+ * https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/10ef9ada-cba2-4384-ab60-7b6290ed4a9a
+ *
+ * @param device
+ *     The device that completed the operation requested by a prior I/O
+ *     request.
+ *
+ * @param completion_id
+ *     The completion ID of the I/O request that requested the operation.
+ *
+ * @param status
+ *     An NTSTATUS code describing the success/failure of the operation that
+ *     was completed.
+ *
+ * @param size
+ *     The number of additional bytes to reserve at the end of the resulting
+ *     stream for additional fields to be appended.
+ *
+ * @return
+ *     A new wStream containing an I/O completion header, followed by the
+ *     requested additional free space.
+ */
+wStream* guac_rdpdr_new_io_completion(guac_rdpdr_device* device,
+        int completion_id, int status, int size);
+
+/**
+ * Initializes device redirection support (file transfer, printing, etc.) for
+ * RDP and handling of the RDPDR channel.  If failures occur, messages noting
```

Comment:
Another double space here

Reply:
Should now be fixed - I did an overall search for double spaces occurring within comments and corrected all occurrences.